components/script/dom/htmlmediaelement.rs

```diff
@@ -1650,7 +1818,138 @@ impl HTMLMediaElementContext {
+    image_cache: Arc<dyn ImageCache>,
+    /// The element that initiated the request.
+    elem: Trusted<HTMLMediaElement>,
+    /// The cache ID for this request.
+    id: PendingImageId,
+    /// Indicates whether the request failed, and why
+    status: Result<(), NetworkError>,
+    /// True if this response is invalid and should be ignored.
+    aborted: bool,
+    /// Timing data for this resource
+    resource_timing: ResourceFetchTiming,
+    /// Url for the resource
+    url: ServoUrl,
+}
+
+impl FetchResponseListener for PosterFrameFetchContext {
+    fn process_request_body(&mut self) {}
+    fn process_request_eof(&mut self) {}
+
+    fn process_response(&mut self, metadata: Result<FetchMetadata, NetworkError>) {
+        self.image_cache
+            .notify_pending_response(self.id, FetchResponseMsg::ProcessResponse(metadata.clone()));
+
+        let metadata = metadata.ok().map(|meta| match meta {
+            FetchMetadata::Unfiltered(m) => m,
+            FetchMetadata::Filtered { unsafe_, .. } => unsafe_,
+        });
+
+        // Step 14.5 of https://html.spec.whatwg.org/multipage/#img-environment-changes
+        if let Some(metadata) = metadata.as_ref() {
+            if let Some(ref content_type) = metadata.content_type {
+                let mime: Mime = content_type.clone().into_inner().into();
+                if mime.type_() == mime::MULTIPART && mime.subtype().as_str() == "x-mixed-replace" {
+                    self.aborted = true;
+                    let elem = self.elem.root();
+                    elem.poster_frame_canceller.borrow_mut().cancel();
+                    elem.poster_generation_id
+                        .set(elem.poster_generation_id.get() + 1);
+                }
+            }
+        }
+
+        let status_code = metadata
+            .as_ref()
+            .and_then(|m| m.status.as_ref().map(|&(code, _)| code))
+            .unwrap_or(0);
+
+        self.status = match status_code {
+            0 => Err(NetworkError::Internal(
+                "No http status code received".to_owned(),
+            )),
+            200...299 => Ok(()),
+            _ => Err(NetworkError::Internal(format!(
+                "HTTP error code {}",
+                status_code
+            ))),
+        };
+    }
+
+    fn process_response_chunk(&mut self, payload: Vec<u8>) {
+        if self.aborted {
+            // An error was received previously, skip processing the payload.
+            return;
+        }
+
+        if self.status.is_ok() {
+            self.image_cache
+                .notify_pending_response(self.id, FetchResponseMsg::ProcessResponseChunk(payload));
+        }
+    }
+
+    fn process_response_eof(&mut self, response: Result<ResourceFetchTiming, NetworkError>) {
+        self.image_cache
+            .notify_pending_response(self.id, FetchResponseMsg::ProcessResponseEOF(response));
+        let elem = self.elem.root();
+        elem.delay_load_event(false);
```

Comment:
I worry that this will cause pages to be considered loaded before the main media loading algorithm says that this is the case.